keps/sig-auth/0034-20190123-harden-default-discovery-bindings.md

```diff
@@ -0,0 +1,91 @@
+
+## Motivation
+
+One of the work items resulting from the [CVE-2018-1002105](https://github.com/kubernetes/kubernetes/issues/71411) post-mortem was to [investigate hardening the default RBAC discovery ClusterRoleBindings](https://github.com/kubernetes/kubernetes/issues/72115) (i.e. `system:discovery` and `system:basic-user`) to limit potential avenues for similar attack. Additionally, the fact that API extensions are exposed by the default discovery bindings is surprising to some and represents a potential privacy concern (e.g. `/apis/self-flying-cars.unicorn.vc/v1/`).
+
+### Goals
+
+Limit the set of APIs which allow for unauthenticated access by default.
+
+### Non-Goals
+
+Disabling unauthenticated and/or anonymous access to all APIs by default. There are several legitimate use-cases for unauthenticated calls, e.g. loadbalancer and kubelet liveliness checks.
+
+## Proposal
+
+* Remove the `system:unauthenticated` subject group from the default [`system:discovery` and `system:basic-user` ClusterRoleBindings](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L526-L527)
+* Create a new default ClusterRole which prominently, explicitly documents its intent to freely disclose information (e.g. `system:public-info-viewer`) and contains the `GET /healthz`, `/version` and `/version/` PolicyRules.
+* Create a new default ClusterRoleBinding for the new ClusterRole, which grants the `system:(un)authenticated` subject groups access.
+
+### Risks and Mitigations
+
+#### Existing customization of `system:discovery`
+
+If a user upgrades a cluster which has modified the `system:discovery` ClusterRoleBinding, these changes could either be trampled or restricted endpoints could end up being re-exposed via the new `system:public-info-viewer` binding.
+
+During the API server's RBAC auto-reconciliation, if `system:discovery` exists and `system:public-info-viewer` does not (i.e. the state after an upgrade) we'll copy `system:discovery`'s subjects and reconciliation annotation value. This will preserve permissions as-is during API server upgrades, even with the addition of `system:public-info-viewer`, since existing customization will be preserved.
+
+#### Dependence on existing unauthenticated behavior
+
+Some use-cases might require the existing permissions to be preserved for unauthenticated calls
+
+In the release notes, we can include an easy 'escape hatch' to re-enable unauthenticated access to the APIs, e.g. `kubectl create clusterrolebinding public-discovery --clusterrole=system:discovery --group=system:unauthenticated`
+
+From [`system:basic-user`](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L209-L215):
```

Comment:
nit: when linking to a line number, use a fixed commit version (press 'y' on github to replace the URL with a canonical version)